posthog/api/special_migration.py

```diff
@@ -0,0 +1,76 @@
+from rest_framework import response, serializers, viewsets
+from rest_framework.decorators import action
+
+from posthog.api.routing import StructuredViewSetMixin
+from posthog.celery import app
+from posthog.models.special_migration import MigrationStatus, SpecialMigration, get_all_running_special_migrations
+from posthog.permissions import StaffUser
+from posthog.special_migrations.runner import process_error
+from posthog.tasks.special_migrations import run_special_migration
+
+# allow users to set this?
+# important to prevent us taking up too many celery workers
+MAX_CONCURRENT_SPECIAL_MIGRATIONS = 1
+
+
+class SpecialMigrationSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = SpecialMigration
+        fields = [
+            "id",
+            "name",
+            "progress",
+            "status",
+            "current_operation_index",
+            "current_query_id",
+            "celery_task_id",
+            "started_at",
+            "finished_at",
+            "error",
+        ]
+        read_only_fields = [
+            "id",
+            "name",
+            "progress",
+            "status",
+            "current_operation_index",
+            "current_query_id",
+            "celery_task_id",
+            "started_at",
+            "finished_at",
+            "error",
+        ]
+
+
+class SpecialMigrationsViewset(StructuredViewSetMixin, viewsets.ModelViewSet):
+    queryset = SpecialMigration.objects.all()
+    permission_classes = [StaffUser]
+    serializer_class = SpecialMigrationSerializer
+
+    @action(methods=["POST"], detail=True)
+    def trigger(self, request, **kwargs):
+        if len(get_all_running_special_migrations()) >= MAX_CONCURRENT_SPECIAL_MIGRATIONS:
+            return response.Response(
+                {
+                    "success": False,
+                    "error": f"No more than {MAX_CONCURRENT_SPECIAL_MIGRATIONS} special migrations can run at once",
+                },
+                status=400,
+            )
+        sm = self.get_object()
+        task = run_special_migration.delay(sm.name)
+        sm.celery_task_id = str(task.id)
+        sm.save()
+        return response.Response({"success": True}, status=200)
+
+    # DANGEROUS! Can cause another task to be lost
```

Comment:
Is there a way to _re-queue migrations_ ? I know we check for migration we've completed against at startup but what about having this as well in a cron? (e.g. every 24 h)

This could be a quick and dirty mitigation for this issue.

Reply:
Yeah could do for this daily.

Also I thought about this and in CH deployments Celery tasks aren't really _essential_ anyway, so this isn't as dangerous.